src/ir/local-graph.h

```diff
@@ -151,23 +152,50 @@ struct LocalGraph {
   // Defined publicly as other utilities need similar data layouts.
   using GetSetsMap = std::unordered_map<LocalGet*, Sets>;
 
-private:
-  Mode mode;
+protected:
   Function* func;
   std::set<Index> SSAIndexes;
 
   // A map of each get to the sets relevant to it. This is mutable so that
-  // getSets() can be const: in eager mode no changes to this are ever made in
-  // getSets(), while in lazy mode any changes are just memoization.
+  // getSets() can be const in LazyLocalGraph (which does memoization, see
+  // below).
   mutable GetSetsMap getSetsMap;
 
-  // The internal implementation of the flow analysis used to compute
-  // getSetsMap.
-  struct LocalGraphFlower;
-  std::unique_ptr<LocalGraphFlower> flower;
+  std::unordered_map<LocalSet*, SetInfluences> setInfluences;
+  std::unordered_map<LocalGet*, GetInfluences> getInfluences;
+};
+
+// The internal implementation of the flow analysis used to compute things. This
+// must be declared in the header so that LazyLocalGraph can declare a unique
+// ptr to it, below.
+struct LocalGraphFlower;
+
+struct LazyLocalGraph : public LocalGraph {
+  LazyLocalGraph(Function* func, Module* module = nullptr);
+  ~LazyLocalGraph();
+
+  const Sets& getSets(LocalGet* get) const {
+    auto iter = getSetsMap.find(get);
+    if (iter == getSetsMap.end()) {
+      // A missing entry means we did not do the computation yet. Do it now.
+      computeGetSets(get);
+      iter = getSetsMap.find(get);
+      assert(iter != getSetsMap.end());
+    }
+    return iter->second;
+  }
+
+  // Override parent APIs that are not yet lazy.
+  void computeSetInfluences() { WASM_UNREACHABLE("TODO: Lazify"); }
+  void computeGetInfluences() { WASM_UNREACHABLE("TODO: Lazify"); }
+  void computeSSAIndexes() { WASM_UNREACHABLE("TODO: Lazify"); }
```

Comment:
Are you planning to make these lazy as well? I was imagining having a separate base class under the assumption that the eager and lazy interfaces would always be different.

Reply:
Yes, the next step is to lazify these. The big speed wins will happen there actually.

Reply:
Got it 👍